<h1>Q7) How many records have a new line?</h1>

In [32]:
import re
import pandas as pd
# Load the uploaded CSV file
file_path = 'Data.csv'
df = pd.read_csv(file_path)

# Display the first few rows and the column names to understand the structure
df.head(), df.columns

(                                                                                                                                                        Data
 0                          Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
 1                          Watch or listen live weekdays at 8:30am MT at ryanjespersen.com. Subscribe via YouTube or your favourite podcast app. #RealTalkRJ
 2                                                                                                                       Chubby And Hot, Always Stir The Pot!
 3                                                                                                                       Chubby And Hot, Always Stir The Pot!
 4  Journalist, publisher of Rebel News — telling the other side of the story. Awarded the Queen’s Diamond Jubilee Medal for advancing freedom of expression.,
 Index(['Data'], dtype='object'))

In [33]:
newline_pattern = r'\n'


In [34]:
mask = df['Data'].str.contains(newline_pattern)
newline_count = mask.sum()
matching_rows = df[mask]


In [35]:
print("✅ Number of records with a newline:", newline_count)
print(matching_rows[['Data']].head())

✅ Number of records with a newline: 1211
                                                                                                                                                   Data
21      Think-tank president. Out-of-touch opinion elite. Proud husband of a teacher. Albertan. 🌈 #LGBTQally (he/him)\nMastodon: @seandunn@hachyderm.io
22      Think-tank president. Out-of-touch opinion elite. Proud husband of a teacher. Albertan. 🌈 #LGBTQally (he/him)\nMastodon: @seandunn@hachyderm.io
45  Traveller, nature lover, photographer. HR Professional by day & MBA Candidate. Calgarian. He/him\nOpinions are my own and do not represent any org.
46  Traveller, nature lover, photographer. HR Professional by day & MBA Candidate. Calgarian. He/him\nOpinions are my own and do not represent any org.
64                        📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time for CPC/UCP trolls\n📌 Treaty 7 settler\n📌 Frankly, my dear, I don't give a fuck


**Factors to consider**
* Multiple newlines \n\n\n	May create tokenization artifacts or unintended paragraph breaks.
* Trailing/leading newlines	Can cause padding issues in prompt formatting.
* HTML <br> used instead	Missed by simple \n regex — should optionally detect.
* Invisibles like \r\n (Windows)	Could lead to false negatives if not normalized.
* Text split mid-sentence	Can affect context chunking or coherence in LLM summarization.


In [36]:
# Define a pattern to match all newline variants
newline_pattern = r'(?:\r\n|\r|\n|<br\s*/?>)'

# Count number of newlines in each record
df['newline_count'] = df['Data'].str.count(newline_pattern, flags=re.IGNORECASE)

# Count total records with at least one newline
has_newline_mask = df['newline_count'] >= 1
total_with_newlines = has_newline_mask.sum()

# Count total records with 2 or more newlines
multi_newline_mask = df['newline_count'] >= 2
multi_newline_count = multi_newline_mask.sum()

# Print results
print(f"✅ Total records with at least one newline: {total_with_newlines}")
print(f"✅ Records with 2 or more newlines: {multi_newline_count}")

# Optional: preview those records
print("\n🔍 Sample of records with 2+ newlines:")
print(df[multi_newline_mask][['Data', 'newline_count']].head())


✅ Total records with at least one newline: 1211
✅ Records with 2 or more newlines: 773

🔍 Sample of records with 2+ newlines:
                                                                                                                                               Data  \
64                    📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time for CPC/UCP trolls\n📌 Treaty 7 settler\n📌 Frankly, my dear, I don't give a fuck   
65                    📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time for CPC/UCP trolls\n📌 Treaty 7 settler\n📌 Frankly, my dear, I don't give a fuck   
85   When I look at the mess that is Alberta under the UCP I know I will #NeverVoteConservative\nTweets are my opinion 🏳️‍🌈\n@CanadianInAB@mstdn.ca   
86   When I look at the mess that is Alberta under the UCP I know I will #NeverVoteConservative\nTweets are my opinion 🏳️‍🌈\n@CanadianInAB@mstdn.ca   
139   Ban semi auto rifles w/detachable magazines and handguns as part of a gun violence harm-reduction strategy\n\nNo 

<h2> 1211 records contain a decimal number </h2>

**Reference List (APA 7):**
OpenAI. (2025). ChatGPT (May 17 version) [Large language model]. https://chat.openai.com/  

First Prompt: "How to use regex to find newlines"

Last Prompt : "count how many records have 2 or more newlines? That would help detect paragraphs or long posts, which often require chunking for LLM input"